### DISCUSSION TITLE
Promise Helper

### DISCUSSION THREAD

**Discussion by acro5piano** [remeda/remeda#968] [Suggest a Function]
Can we handle Promise inside `pipe` function? Ramda has such functionalities.

https://ramdajs.com/docs/#pipeP

**Comment by acro5piano** [comment]
It would be great if we can resolve promises like this.

```ts
const res = await R.pipe(
  fetch('https://jsonplaceholder.typicode.com/todos/1'), 
  R.resolve(),
  R.pick(['id', 'title']),
)
console.log(res) // => { id: 1, title: "delectus aut autem" } 
```

where R.resolve is 

```ts
function Resolve<T> = <R>(args: Promise<T>, fn: (value: T) => R): Promise<R>
function Resolve<T> = <R>(fn: (value: T) => R): Promise<R>
```

Is it technically possible and acceptable? If so, I would try to implement it.

Thanks

**Comment by acro5piano** [comment]
I've end up to create my own `andThen` function.

```ts
export function andThen<T, R>(
  fn: (a: T extends Promise<infer S> ? S : never) => R,
): (a: T) => Promise<R> {
  return async (a: T) => {
    return (fn as any)(await a)
  }
}
```

Usage:

```ts
const res = await R.pipe(
  fetch('https://jsonplaceholder.typicode.com/todos'),
  andThen(r => r.json()),
  andThen(
    R.createPipe(
      R.filter(todo => todo.id < 5),
      R.take(3),
      R.map(R.pick(['id', 'title'])),
    ),
  ),
)
console.log(res)

// output:
// [
//   { id: 1, title: 'delectus aut autem' },
//   { id: 2, title: 'quis ut nam facilis et officia qui' },
//   { id: 3, title: 'fugiat veniam minus' }
// ]

```

**Comment by cassus** [comment]
Note to self:

```ts
const res = await 
  fetch('https://jsonplaceholder.typicode.com/todos')
  .then(r => r.json()),
  .then(
    R.createPipe(
      R.filter(todo => todo.id < 5),
      R.take(3),
      R.map(R.pick(['id', 'title'])),
    ),
  ),
)
```

**Comment by eranhirsch** [comment]
You can always mix awaits and remeda, there's no reason to wrap it with a remeda function in my opinion.

```ts
const x = pipe(
  await pipe(
    // non-async functions, e.g.
    map(...),
    filter(...),
    myAsyncFunction, // returns Promise<...>
  ),
  // post async non-async functions, e.g.
  map(...),
  filter(...),
);
```

**Comment by somnicattus** [comment]
Hello.

Remeda cannot handle itarative `Promise`s with lazy evaluation. We need `AsyncIterator` for the functionality.

I published remeda-compatible npm package rotery, which expands remeda with Iterators and AsyncIterators support.

https://github.com/somnicattus/rotery

**Comment by thelinuxlich** [comment]
this is cool but I think it should be merged to Remeda

**Comment by somnicattus** [comment]
@thelinuxlich Actually, I have no idea how the iterator/asyncIterator features should be merged into remeda.

Not many people need iterator/asyncIterator, so it can be even annoying because these features have the same names as some array utilities.
So, please feel free to propose methods for merging the iterator/asyncIterator feature and make PRs.

**Comment by sudall** [comment]
For anyone else struggling with this, https://github.com/sindresorhus/p-pipe seems to work well for me.
